packages/financial-templates-lib/src/helpers/GasEstimator.ts

```diff
@@ -143,12 +152,32 @@ export class GasEstimator {
     } else return { gasPrice: Math.ceil(this.lastFastPriceGwei * 1e9) };
   }
 
+  // Returns the base fee observed the last time the estimator was updated.
+  getCurrentBaseFee(): number {
+    return this.latestBaseFee;
+  }
+
+  getExpectedCumulativeGasPrice(): number {
+    if (this.type == NetworkType.London) return this.latestBaseFee + this.latestMaxPriorityFeePerGasGwei;
+    else return this.latestBaseFee;
+  }
+
   async _update() {
-    const latestGasInfo = await this._getPrice(this.networkId);
+    // Fetch the gasPrice info from the external provider and fetch the base fee from the web3 instance. Note that the
+    // gas price provider on mainnet (etherchain) does provide a base fee but in practice it seems to lag behind the
+    // most recent blocks somewhat. Using the web3 provider is more direct. Also Note that if this is not on L1 then the
+    // web3.eth.getGasPrice call will return the current gas price, not base fee. finally, note that if web3 was not
+    // defined then the base fee will be 0.
+    const [gasInfo, baseFee] = await Promise.all([
+      this._getPrice(this.networkId),
+      this.web3 ? this.web3.eth.getGasPrice() : null, // If a web3 instance was provided, fetch the base fee.
+    ]);
+    this.latestBaseFee = Number(baseFee);
```

Comment:
What if `baseFee == null`? Does this set `this.latestBaseFee = 0` and is that a problem?

Reply:
yup. it would be null. in that case base fee would indeed be 0. I can change the `getCurrentBaseFee` and `getExpectedCumulativeGasPrice` to return an error if it's 0, wdyt?

Reply:
Is it ok to return 0 as a base fee? Otherwise we should throw an error yes

Reply:
well no. if it is not london then `web3.eth. getGasPrice` returns the most recent gas price.

Reply:
Ahhh, that makes sense. My bad. 

What do you think about having an `if (this.type == NetworkType.London)` at the top of this method and making different calls and setting different variables depending on the type. It's a little confusing (IMO) to overload a variable called latestBaseFee to store the average gas price for non-london chains.

Then we could maybe (?) just have an interface that has methods for:
1. An estimate of the gas price that you will actually pay.
2. The max gas price you will pay.
3. The suggested transaction params that you should pass to web3.

These could all be computed differently depending on the type, but regardless of london or not, they should be computable, right? Note: this could be already what you're doing and I could just be misunderstanding. Thoughts?

Reply:
ok so I've got for a minimal implementation that achieves what you laid out without needing to make big changes to the GasEstimator and while considering the recommendations you had in https://github.com/UMAprotocol/protocol/pull/3656#discussion_r761994730

now, on `update`, we fetch the most "latest" block, if there is a web3 and this is london. Then, if london, we extract the `latestBaseFee` and store it.

When calling `getExpectedCumulativeGasPrice` it returns:
1) `latestBaseFee + latestMaxPriorityFeePerGasGwei` if london and
2) `lastFastPriceGwei` if we are on any non-london chain.

`lastFastPriceGwei` seems to be the most accurate version of the expected amount actually paid if it's not london as this is the prevailing gas price the bot will pay. no additional case work is needed.

Please let me know if you agree with this! I really want to keep this as simple as possible and I feel this archives this.